CorruptIndexException will be thrown if tablet server failover

### Search before asking

- [x] I searched in the [issues](https://github.com/alibaba/fluss/issues) and found nothing similar.


### Fluss version

main (development)

### Please describe the bug 🐞

Kill the tabletserver, and restart it.. The following exception will be thrown:
```
2025-03-07 14:02:04,265 ERROR com.alibaba.fluss.server.ServerBase                          [] - Could not start the TabletServer.
com.alibaba.fluss.exception.CorruptIndexException: Corrupt time index found, time index file (/mnt/fluss/data/5/remote-log-index-cache/2063501_8fbacbc5-6fdb-4b37-9473-1b697d719cee.timeindex) has non-zero size but the last timestamp is 0 which is less than the first timestamp 1741224500729
2025-03-07 14:02:04,271 ERROR com.alibaba.fluss.server.ServerBase                          [] - Could not start TabletServer.
com.alibaba.fluss.exception.FlussException: Failed to start the TabletServer.
        at com.alibaba.fluss.server.ServerBase.start(ServerBase.java:131) ~[fluss-server-0.6-SNAPSHOT.jar:0.6-SNAPSHOT]
        at com.alibaba.fluss.server.ServerBase.startServer(ServerBase.java:88) [fluss-server-0.6-SNAPSHOT.jar:0.6-SNAPSHOT]
        at com.alibaba.fluss.server.tablet.TabletServer.main(TabletServer.java:135) [fluss-server-0.6-SNAPSHOT.jar:0.6-SNAPSHOT]
Caused by: com.alibaba.fluss.exception.CorruptIndexException: Corrupt time index found, time index file (/mnt/fluss/data/5/remote-log-index-cache/
```

### Solution

The timeindex does non-zero size, but the last entry read is 0.. Maybe tabletserver is downloading the timeindex, while fail directly. So the timeindex is not completed which cause the exception..
Maybe we can delete the corrupted timeindex and enable it to be download again insteading throwing exceptio directly.
Found a similar issue in Kafka https://issues.apache.org/jira/browse/KAFKA-2012

### Are you willing to submit a PR?

- [x] I'm willing to submit a PR!

--- a/fluss-server/src/main/java/com/alibaba/fluss/server/exception/CorruptIndexException.java
+++ b/fluss-server/src/main/java/com/alibaba/fluss/server/exception/CorruptIndexException.java
@@ -5,7 +5,7 @@
  * you may not use this file except in compliance with the License.
  * You may obtain a copy of the License at
  *
- *      http://www.apache.org/licenses/LICENSE-2.0
+ *     http://www.apache.org/licenses/LICENSE-2.0
  *
  * Unless required by applicable law or agreed to in writing, software
  * distributed under the License is distributed on an "AS IS" BASIS,
@@ -14,7 +14,7 @@
  * limitations under the License.
  */
 
-package com.alibaba.fluss.exception;
+package com.alibaba.fluss.server.exception;
 
 import com.alibaba.fluss.annotation.PublicEvolving;
 
@@ -25,7 +25,7 @@ import com.alibaba.fluss.annotation.PublicEvolving;
  * @since 0.1
  */
 @PublicEvolving
-public class CorruptIndexException extends RetriableException {
+public class CorruptIndexException extends RuntimeException {
     public CorruptIndexException(String message) {
         super(message);
     }

--- a/fluss-server/src/main/java/com/alibaba/fluss/server/log/AbstractIndex.java
+++ b/fluss-server/src/main/java/com/alibaba/fluss/server/log/AbstractIndex.java
@@ -16,8 +16,8 @@
 
 package com.alibaba.fluss.server.log;
 
-import com.alibaba.fluss.exception.CorruptIndexException;
 import com.alibaba.fluss.exception.IndexOffsetOverflowException;
+import com.alibaba.fluss.server.exception.CorruptIndexException;
 import com.alibaba.fluss.utils.FileUtils;
 import com.alibaba.fluss.utils.IOUtils;
 import com.alibaba.fluss.utils.OperatingSystem;

--- a/fluss-server/src/main/java/com/alibaba/fluss/server/log/OffsetIndex.java
+++ b/fluss-server/src/main/java/com/alibaba/fluss/server/log/OffsetIndex.java
@@ -16,9 +16,9 @@
 
 package com.alibaba.fluss.server.log;
 
-import com.alibaba.fluss.exception.CorruptIndexException;
 import com.alibaba.fluss.exception.IndexOffsetOverflowException;
 import com.alibaba.fluss.exception.InvalidOffsetException;
+import com.alibaba.fluss.server.exception.CorruptIndexException;
 
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;

--- a/fluss-server/src/main/java/com/alibaba/fluss/server/log/TimeIndex.java
+++ b/fluss-server/src/main/java/com/alibaba/fluss/server/log/TimeIndex.java
@@ -16,8 +16,8 @@
 
 package com.alibaba.fluss.server.log;
 
-import com.alibaba.fluss.exception.CorruptIndexException;
 import com.alibaba.fluss.exception.InvalidOffsetException;
+import com.alibaba.fluss.server.exception.CorruptIndexException;
 
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;

--- a/fluss-server/src/main/java/com/alibaba/fluss/server/log/remote/RemoteLogIndexCache.java
+++ b/fluss-server/src/main/java/com/alibaba/fluss/server/log/remote/RemoteLogIndexCache.java
@@ -406,16 +406,27 @@ public class RemoteLogIndexCache implements Closeable {
                     if (Files.exists(offsetIndexFile.toPath())
                             && Files.exists(timestampIndexFile.toPath())) {
                         long offset = offsetFromRemoteIndexCacheFileName(indexFileName);
-                        OffsetIndex offsetIndex =
-                                new OffsetIndex(offsetIndexFile, offset, Integer.MAX_VALUE, false);
-                        offsetIndex.sanityCheck();
+                        try {
+                            OffsetIndex offsetIndex =
+                                    new OffsetIndex(
+                                            offsetIndexFile, offset, Integer.MAX_VALUE, false);
+                            offsetIndex.sanityCheck();
 
-                        TimeIndex timeIndex =
-                                new TimeIndex(timestampIndexFile, offset, Integer.MAX_VALUE, false);
-                        timeIndex.sanityCheck();
+                            TimeIndex timeIndex =
+                                    new TimeIndex(
+                                            timestampIndexFile, offset, Integer.MAX_VALUE, false);
+                            timeIndex.sanityCheck();
 
-                        Entry entry = new Entry(offsetIndex, timeIndex);
-                        internalCache.put(remoteSegmentId, entry);
+                            Entry entry = new Entry(offsetIndex, timeIndex);
+                            internalCache.put(remoteSegmentId, entry);
+                        } catch (CorruptIndexException e) {
+                            LOG.debug(
+                                    "Remote offset/time log index is corrupt, delete corrupt index.",
+                                    e);
+                            // let's delete offset index & time index
+                            Files.deleteIfExists(offsetIndexFile.toPath());
+                            Files.deleteIfExists(timestampIndexFile.toPath());
+                        }
                     } else {
                         // Delete all of them if any one of those indexes is not available for a
                         // specific segment id.

--- a/fluss-server/src/main/java/com/alibaba/fluss/server/log/remote/RemoteLogIndexCache.java
+++ b/fluss-server/src/main/java/com/alibaba/fluss/server/log/remote/RemoteLogIndexCache.java
@@ -17,10 +17,10 @@
 package com.alibaba.fluss.server.log.remote;
 
 import com.alibaba.fluss.annotation.VisibleForTesting;
-import com.alibaba.fluss.exception.CorruptIndexException;
 import com.alibaba.fluss.exception.FlussRuntimeException;
 import com.alibaba.fluss.exception.RemoteStorageException;
 import com.alibaba.fluss.remote.RemoteLogSegment;
+import com.alibaba.fluss.server.exception.CorruptIndexException;
 import com.alibaba.fluss.server.log.OffsetIndex;
 import com.alibaba.fluss.server.log.OffsetPosition;
 import com.alibaba.fluss.server.log.StorageAction;

--- a/fluss-server/src/test/java/com/alibaba/fluss/server/log/TimeIndexTest.java
+++ b/fluss-server/src/test/java/com/alibaba/fluss/server/log/TimeIndexTest.java
@@ -16,8 +16,8 @@
 
 package com.alibaba.fluss.server.log;
 
-import com.alibaba.fluss.exception.CorruptIndexException;
 import com.alibaba.fluss.exception.InvalidOffsetException;
+import com.alibaba.fluss.server.exception.CorruptIndexException;
 
 import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Test;

--- a/fluss-server/src/test/java/com/alibaba/fluss/server/log/remote/RemoteLogIndexCacheTest.java
+++ b/fluss-server/src/test/java/com/alibaba/fluss/server/log/remote/RemoteLogIndexCacheTest.java
@@ -31,7 +31,10 @@ import org.junit.jupiter.params.ParameterizedTest;
 import org.junit.jupiter.params.provider.ValueSource;
 
 import java.io.File;
+import java.nio.ByteBuffer;
+import java.nio.channels.FileChannel;
 import java.nio.file.Path;
+import java.nio.file.StandardOpenOption;
 import java.util.Arrays;
 import java.util.UUID;
 
@@ -96,6 +99,34 @@ class RemoteLogIndexCacheTest extends RemoteLogTestBase {
         LogTablet logTablet = makeLogTabletAndAddSegments(partitionTable);
         // 1. first upload one segment to remote.
         RemoteLogSegment remoteLogSegment = copyLogSegmentToRemote(logTablet, remoteLogStorage, 0);
+        TimeIndex timeIndex = rlIndexCache.getIndexEntry(remoteLogSegment).timeIndex();
+        TimestampOffset timestampOffset = timeIndex.entry(0);
+
+        OffsetIndex offsetIndex = rlIndexCache.getIndexEntry(remoteLogSegment).offsetIndex();
+        OffsetPosition offsetPosition = offsetIndex.lookup(timestampOffset.offset);
+        long resultOffset =
+                rlIndexCache.lookupOffsetForTimestamp(remoteLogSegment, timestampOffset.timestamp);
+        assertThat(offsetPosition.getOffset()).isEqualTo(resultOffset);
+    }
+
+    @ParameterizedTest
+    @ValueSource(booleans = {true, false})
+    void testInitWithCorruptIndex(boolean partitionTable) throws Exception {
+        LogTablet logTablet = makeLogTabletAndAddSegments(partitionTable);
+        // 1. first upload one segment to remote.
+        RemoteLogSegment remoteLogSegment = copyLogSegmentToRemote(logTablet, remoteLogStorage, 0);
+
+        rlIndexCache = new RemoteLogIndexCache(1024 * 1024L, remoteLogStorage, tempDir);
+        File file = rlIndexCache.getIndexEntry(remoteLogSegment).timeIndex().file();
+        // mock corrupt index
+        try (FileChannel fileChannel = FileChannel.open(file.toPath(), StandardOpenOption.APPEND)) {
+            for (int i = 0; i < 12; i++) {
+                fileChannel.write(ByteBuffer.wrap(new byte[] {0}));
+            }
+        }
+
+        // re-initialize index cache and lookup offset
+        rlIndexCache = new RemoteLogIndexCache(1024 * 1024L, remoteLogStorage, tempDir);
         TimeIndex timeIndex = rlIndexCache.getIndexEntry(remoteLogSegment).timeIndex();
         TimestampOffset timestampOffset = timeIndex.entry(0);
 

